### PR TITLE
Fix derivation path checks

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
Fixes https://app.clickup.com/t/9002019994/CHI-4582

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [CHI-4582 Correct xpub/address derivation path checks on Device](https://app.clickup.com/t/9002019994/CHI-4582)

File: common/coin_support/coin_utils.h [hunk 1]
```diff
@@ -29,6 +29,12 @@
 #include "sha2.h"
 #include "utils.h"
 
+#define IS_HARDENED(x) ((x & 0x80000000) == 0x80000000)
+#define IS_NON_HARDENED(x) ((x & 0x80000000) == 0)
```
Comment (TejasvCypherock) [inline common/coin_support/coin_utils.h:33]: We can convert this to inline functions to enforce type of x

File: common/coin_support/coin_utils.h [hunk 2]
```diff
@@ -29,6 +29,12 @@
...
+
+#define XPUB_DEFAULT_DEPTH 3
+#define ADDR_DEFAULT_DEPTH 5
```
Comment (TejasvCypherock) [inline common/coin_support/coin_utils.h:36]: Add comments explaining why/how these are chosen or rename to make them clear

File: common/coin_support/solana.c [hunk 3]
```diff
@@ -284,14 +284,17 @@ bool sol_verify_derivation_path(const uint32_t *path, uint8_t levels) {
...
+          (purpose == NON_SEGWIT && coin == SOLANA && IS_HARDENED(account));
+    } break;
 
     case 4: {    // m/44'/501'/i'/0'
       uint32_t change = path[3];
-      status =
-          (purpose == NON_SEGWIT && coin == SOLANA && change == 0x80000000);
+      uint32_t account = path[2];
+      status = (purpose == NON_SEGWIT && coin == SOLANA &&
+                IS_HARDENED(account) && change == 0x80000000);
```
Comment (TejasvCypherock) [inline common/coin_support/solana.c:297]: Can we add 0x80000000 as a macro in solana.h

**Review by TejasvCypherock — CHANGES_REQUESTED** [review]

**Review by TejasvCypherock — APPROVED** [review]

**Pull request merged**